// @meeco/keystore-api-sdk/lib/configuration.js
"use strict";
// tslint:disable
/**
 * Keystore
 * Storage for secrets and keys. When a user creates and encrypts Data Encryption Keys and Keypairs, they are stored in the Keystore. A user cannot access the Vault without first using the Keystore. In the current implementation the Keystore functionality is implemented by meeco-keystore.
 *
 * OpenAPI spec version: 2.0.0
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var Configuration = /** @class */ (function () {
    function Configuration(param) {
        if (param === void 0) { param = {}; }
        this.apiKey = param.apiKey;
        this.username = param.username;
        this.password = param.password;
        this.accessToken = param.accessToken;
        this.basePath = param.basePath;
    }
    return Configuration;
}());
exports.Configuration = Configuration;
