// @meeco/cryppo/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var buffer_1 = require("buffer");
if (typeof window !== 'undefined' && typeof window.global === 'undefined') {
    // Ensures browser will run without manual polyfills in Angular
    window.Buffer = buffer_1.Buffer;
    window.global = window;
}
__exportStar(require("./decryption/decryption"), exports);
__exportStar(require("./encryption/encryption"), exports);
__exportStar(require("./key-derivation/derived-key"), exports);
__exportStar(require("./key-derivation/pbkdf2-hmac"), exports);
__exportStar(require("./key-pairs/rsa"), exports);
__exportStar(require("./signing/rsa-signature"), exports);
__exportStar(require("./strategies"), exports);
var util_1 = require("./util");
Object.defineProperty(exports, "binaryBufferToString", { enumerable: true, get: function () { return util_1.binaryBufferToString; } });
Object.defineProperty(exports, "decodeDerivationArtifacts", { enumerable: true, get: function () { return util_1.decodeDerivationArtifacts; } });
Object.defineProperty(exports, "decodeSafe64", { enumerable: true, get: function () { return util_1.decodeSafe64; } });
Object.defineProperty(exports, "deSerializeDerivedKeyOptions", { enumerable: true, get: function () { return util_1.deSerializeDerivedKeyOptions; } });
Object.defineProperty(exports, "encodeDerivationArtifacts", { enumerable: true, get: function () { return util_1.encodeDerivationArtifacts; } });
Object.defineProperty(exports, "encodeSafe64", { enumerable: true, get: function () { return util_1.encodeSafe64; } });
Object.defineProperty(exports, "generateEncryptionVerificationArtifacts", { enumerable: true, get: function () { return util_1.generateEncryptionVerificationArtifacts; } });
Object.defineProperty(exports, "generateRandomKey", { enumerable: true, get: function () { return util_1.generateRandomKey; } });
Object.defineProperty(exports, "stringAsBinaryBuffer", { enumerable: true, get: function () { return util_1.stringAsBinaryBuffer; } });
